kafka-elastic.go

```diff
@@ -0,0 +1,341 @@
+		}
+	}()
+
+	// consume notifications
+	ke.wg.Add(1)
+	shutdown := make(chan bool)
+	ke.consumerShutdowns = append(ke.consumerShutdowns, shutdown)
+	go func() {
+		defer ke.wg.Done()
+		for ntf := range consumer.Notifications() {
+			logger.Info(fmt.Sprintf("Rebalanced: %+v\n", ntf))
+		}
+	}()
+
+	// Consume messages
+	ke.wg.Add(1)
+	go func() {
+		defer ke.wg.Done()
+		logKafka.Info("Listening for messages on topic: " + topic)
+		for {
+			select {
+			case kmessage, ok := <-consumer.Messages():
+				if ok {
+					doc := KafkaElasticDoc{
+						Offset:         kmessage.Offset,
+						Value:          "[replace]",
+						TimestampNano:  kmessage.Timestamp.UnixNano(),
+						TimestampPST:   kmessage.Timestamp.Format(time.RFC1123Z),
+						IndexTimestamp: time.Now().UnixNano(),
+						Partition:      kmessage.Partition,
+						Key:            string(kmessage.Key),
+						ValueRaw:       string(kmessage.Value),
+						Topic:          topic,
+					}
+					ke.indexQueue <- &doc
+					consumer.MarkOffset(kmessage, "")
+				}
+			case <-shutdown:
+				logger.Info("Gracefully closing the " + topic + " consumer")
+				consumer.Close()
+				return
+
+			}
+		}
+	}()
+	return nil
+}
+
+func (ke *KafkaElastic) startIndexQueue() error {
+	ke.wg.Add(1)
+
+	go func() {
+		logElastic.Info("Starting indexer...")
+		bulkRequest := ke.elasticClient.Bulk()
+		ticker := time.NewTicker(time.Second * 5)
+
+		defer ke.wg.Done()
+		for {
+			select {
+			case item := <-ke.indexQueue:
+				jsonObj, err := json.Marshal(item)
+				if err != nil {
+					logElastic.Error(err)
+					continue
+				}
+				value := item.ValueRaw
+
+				// Number's are valid json, but need to be treated as raw value
+				_, intErr := strconv.ParseInt(item.ValueRaw, 10, 64)
+				if !json.Valid([]byte(item.ValueRaw)) || intErr == nil {
+					value = `{"value-raw": "` + value + `"}`
+				}
+
+				jsonString := strings.Replace(string(jsonObj), "\"[replace]\"", value, 1)
+
+				bulkRequest.Add(elastic.NewBulkIndexRequest().
+					Index(item.Topic).
+					Id(strconv.FormatInt(item.Offset, 10)).
+					Doc(jsonString))
+				if bulkRequest.NumberOfActions() > 1000 {
+					// Do sends the bulk requests to Elasticsearch
+					_, err := bulkRequest.Do(context.Background())
+					if err != nil {
+						logElastic.Error(err)
+					} else {
+						logElastic.Info("Sent 1000 messages to be index")
+					}
+				}
+			case <-ticker.C:
+				if bulkRequest.NumberOfActions() > 0 {
+					// Do sends the bulk requests to Elasticsearch
+					total := strconv.Itoa(bulkRequest.NumberOfActions())
+
+					_, err := bulkRequest.Do(context.Background())
+					if err != nil {
+						logElastic.Error(err)
+					} else {
+						logElastic.Info("Sent " + total + " messages to be index")
+					}
+				}
+			case <-ke.indexQueueShutdown:
+				return
+			}
+		}
+	}()
+
+	return nil
+}
+
+// Start boot up the kafka elastic daemon process
+func (ke *KafkaElastic) Start() error {
+	err := ke.RefreshTopicIndexes()
```

Comment:
Is RefreshTopicIndexes only ever run once on startup?  If so, would it ever make sense to install as one of your signal handlers (e.g., for USR1 or HUP) something to re-run RefreshTopicIndexes?